# Práctica 1.1 — Probabilidad y Estadística
## Análisis de la evolución poblacional mundial
### Juan Carlos Gallegos Gutiérrez  -   ISSC 512    -   17/08/2026

### Objetivo
Utilizar un conjunto de datos real para comprobar el funcionamiento del entorno de análisis de datos y aplicar medidas estadísticas descriptivas básicas utilizando Pandas y NumPy.

### Actividad 1 — Identificación de la fuente

#### Cita de la fuente
Los datos utilizados en esta práctica provienen de **Our World in Data (OWID)**.
La fuente original es **United Nations, World Population Prospects (2024)**.

**Cita:**
> UN, World Population Prospects (2024) – processed by Our World in Data. “Annual change in population – UN WPP” [dataset]. United Nations, “World Population Prospects”; United Nations, “World Population Prospects - Interim Update” [original data].

#### Preguntas
1. ¿Quién produjo originalmente los datos?
**Respuesta:** United Nations, World Population Prospects (2024).

2. ¿Qué organización los procesa para su utilización en Our World in Data?
**Respuesta:** Our World in Data.

3. ¿Qué representa la variable que analizarás?
**Respuesta:** Representa el cambio anual en la población (crecimiento o decrecimiento poblacional en un año determinado).

### Actividad 2 — Carga y exploración de datos

#### 2.1 Importar las bibliotecas

In [27]:
import pandas as pd
import numpy as np

#### 2.2 Cargar el archivo CSV

In [28]:
df = pd.read_csv('../data/raw/annual-population-growth/annual-population-growth.csv')

#### 2.3 Explorar el DataFrame

In [29]:
df.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN
4,Afghanistan,AFG,1955,119027.0,NaN


In [30]:
df.shape

(38400, 5)

In [31]:
df.columns.tolist()

['Entity',
 'Code',
 'Year',
 'Annual change in population',
 'Annual population change (Projected)']

In [32]:
df.dtypes

Entity                                      str
Code                                        str
Year                                      int64
Annual change in population             float64
Annual population change (Projected)    float64
dtype: object

In [33]:
df.describe()

,Year,Annual change in population,Annual population change (Projected)
count,38400.000000,1.868800e+04,1.971200e+04
mean,2025.500000,2.063019e+06,8.375481e+05
std,43.300872,9.440859e+06,5.816745e+06
min,1951.000000,-3.315935e+06,-2.291035e+07
25%,1988.000000,1.781000e+03,-9.742750e+03
50%,2025.500000,4.689800e+04,9.350000e+01
75%,2063.000000,3.021128e+05,1.411160e+05
max,2100.000000,9.337137e+07,7.206736e+07


In [34]:
df.isnull().sum()

Entity                                      0
Code                                     1200
Year                                        0
Annual change in population             19712
Annual population change (Projected)    18688
dtype: int64

In [35]:
df["Entity"].nunique()

256

#### Preguntas
1. ¿Cuántas filas y columnas tiene el dataset?
**Respuesta:** El dataset tiene 38,400 filas y 5 columnas (`df.shape`).

2. ¿Qué tipo de datos contiene la variable `Year`?
**Respuesta:** Entero (`int64`) (`df.dtypes`).

3. ¿Existen valores faltantes?
**Respuesta:** Sí, existen 1,200 valores nulos en `Code`, 19,712 en `Annual change in population`, y 18,688 en `Annual population change (Projected)` (`df.isnull().sum()`).

4. ¿Cuántas entidades diferentes contiene el dataset?
**Respuesta:** Contiene 256 entidades diferentes (`df["Entity"].nunique()`).

### Actividad 3 — Estadística descriptiva

In [36]:
df_2023 = df[df["Year"] == 2023]
cambio_anual = df_2023["Annual change in population"]

In [37]:
cambio_anual.mean()

np.float64(2008097.765625)

In [38]:
cambio_anual.median()

np.float64(46865.5)

In [39]:
cambio_anual.var()

np.float64(82427560967665.19)

In [40]:
cambio_anual.std()

np.float64(9078962.549083743)

In [41]:
cambio_anual.min()

np.float64(-3315935.0)

In [42]:
cambio_anual.max()

np.float64(72315107.0)

#### Preguntas
1. ¿Cuál es el valor de la media?
**Respuesta:** 2,008,097.77 (`cambio_anual.mean()`).

2. ¿Cuál es el valor de la mediana?
**Respuesta:** 46,865.5 (`cambio_anual.median()`).

3. ¿Cuál es el valor de la desviación estándar?
**Respuesta:** 9,078,962.55 (`cambio_anual.std()`).

4. ¿Qué entidad presenta el valor mínimo y cuál el máximo?
**Respuesta:** El valor mínimo es -3,315,935.0 (Ukraine) y el valor máximo es 72,315,107.0 (Less developed regions, excluding China) (calculado con `cambio_anual.min()` y `cambio_anual.max()` y filtrando en `df_2023`).

In [43]:
entidad_min = df_2023[df_2023["Annual change in population"] == cambio_anual.min()]["Entity"].values[0]
entidad_max = df_2023[df_2023["Annual change in population"] == cambio_anual.max()]["Entity"].values[0]
print(entidad_min)
print(entidad_max)

Ukraine
Less developed regions, excluding China


### Actividad 4 — Comprobación del entorno

In [44]:
import sys
print("Python:", sys.version)
print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)

Python: 3.14.3 (tags/v3.14.3:323c59a, Feb  3 2026, 16:04:56) [MSC v.1944 64 bit (AMD64)]
Pandas: 3.0.5
NumPy: 2.5.2


### Países que inician con G y estadísticas de Germany

Se identifican las entidades cuyo nombre comienza con la letra **G** (Gallegos) y se calculan las estadísticas descriptivas del cambio anual de población de **Germany**.

In [45]:
# Países o entidades que inician con la letra G
paises_g = df.loc[df["Entity"].str.startswith("G", na=False), "Entity"].unique()
print("Países o entidades que inician con G:")
print(paises_g)

Países o entidades que inician con G:
<StringArray>
[        'Gabon',        'Gambia',       'Georgia',       'Germany',
         'Ghana',     'Gibraltar',        'Greece',     'Greenland',
       'Grenada',    'Guadeloupe',          'Guam',     'Guatemala',
      'Guernsey',        'Guinea', 'Guinea-Bissau',        'Guyana']
Length: 16, dtype: str


In [46]:
# Cambio anual de población histórico de Germany
germany = df.loc[
    (df["Entity"] == "Germany") &
    (df["Annual change in population"].notna()),
    "Annual change in population"
]

# Estadísticas descriptivas
print(germany.describe())
print("Varianza:", germany.var())

count        73.000000
mean     201376.095890
std      307378.667262
min     -403425.000000
25%       26961.000000
50%      167475.000000
75%      462006.000000
max      701468.000000
Name: Annual change in population, dtype: float64
Varianza: 94481645088.00456


### Países con cambio anual mínimo y máximo

Se identifican las entidades y los años correspondientes a los valores mínimo y máximo de `Annual change in population` de todo el DataFrame.

In [47]:
columna_cambio = "Annual change in population"

# idxmin e idxmax devuelven los índices de los valores extremos, ignorando NaN
fila_min = df.loc[df[columna_cambio].idxmin(), ["Entity", "Year", columna_cambio]]
fila_max = df.loc[df[columna_cambio].idxmax(), ["Entity", "Year", columna_cambio]]

print("Cambio anual mínimo:")
print(fila_min)
print("\nCambio anual máximo:")
print(fila_max)

Cambio anual mínimo:
Entity                           Ukraine
Year                                2023
Annual change in population   -3315935.0
Name: 35772, dtype: object

Cambio anual máximo:
Entity                              World
Year                                 1990
Annual change in population    93371370.0
Name: 37839, dtype: object


In [48]:
df.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN
4,Afghanistan,AFG,1955,119027.0,NaN


In [49]:
'''
loc busca por etiquetas, iloc busca por posición.
[filas, columnas]

 0 1 2 3 ...
0
1
2
3
...
'''
df.loc[:,"Entity"]
df.iloc[:,0:2]
df.loc[600:610, ["Entity", "Year"]]

,Entity,Year
600,American Samoa,1951
601,American Samoa,1952
602,American Samoa,1953
603,American Samoa,1954
604,American Samoa,1955
605,American Samoa,1956
606,American Samoa,1957
607,American Samoa,1958
608,American Samoa,1959
609,American Samoa,1960


In [50]:
'''Ejercicio 1'''
print(df.loc[10000])
'''Ejercicio 2'''
print(df.loc[:, ["Entity", "Year", "Annual change in population"]])
'''Ejercicio 3'''
print(df.loc[10000, ["Entity", "Year", "Annual change in population"]])
'''Ejercicio 4'''
print(df.loc[(df["Entity"] == "Mexico") & (df["Year"] == 2020), :])
'''Ejercicio 5'''
print(df.iloc[0:5])
'''Ejercicio 6'''
print(df.iloc[:, [0, 2, 3]])
'''Ejercicio 7'''
print(df.iloc[0:4, 0:3])

Entity                                  Equatorial Guinea
Code                                                  GNQ
Year                                                 2051
Annual change in population                           NaN
Annual population change (Projected)              46407.0
Name: 10000, dtype: object
            Entity  Year  Annual change in population
0      Afghanistan  1951                     103163.0
1      Afghanistan  1952                     108441.0
2      Afghanistan  1953                     108919.0
3      Afghanistan  1954                     111251.0
4      Afghanistan  1955                     119027.0
...            ...   ...                          ...
38395     Zimbabwe  2096                          NaN
38396     Zimbabwe  2097                          NaN
38397     Zimbabwe  2098                          NaN
38398     Zimbabwe  2099                          NaN
38399     Zimbabwe  2100                          NaN

[38400 rows x 3 columns]
Entity   

In [51]:
df.set_index("Entity")

,Code,Year,Annual change in population,Annual population change (Projected)
Entity,,,,
Afghanistan,AFG,1951,103163.0,NaN
Afghanistan,AFG,1952,108441.0,NaN
Afghanistan,AFG,1953,108919.0,NaN
Afghanistan,AFG,1954,111251.0,NaN
Afghanistan,AFG,1955,119027.0,NaN
...,...,...,...,...
Zimbabwe,ZWE,2096,NaN,100752.0
Zimbabwe,ZWE,2097,NaN,91796.0
Zimbabwe,ZWE,2098,NaN,87605.0


In [53]:
df.head()

,Entity,Code,Year,Annual change in population,Annual population change (Projected)
0,Afghanistan,AFG,1951,103163.0,NaN
1,Afghanistan,AFG,1952,108441.0,NaN
2,Afghanistan,AFG,1953,108919.0,NaN
3,Afghanistan,AFG,1954,111251.0,NaN
4,Afghanistan,AFG,1955,119027.0,NaN
